### ISSUE TITLE
Center Text

### ISSUE THREAD

**Issue by DasLixou** [grovesNL/glyphon#54]
How would I center the text in for example the middle of my window?

I currently do
```rust
let x = self.size.width / 2;
        let y = self.size.height / 2;

        self.text_renderer
            .prepare(
                &self.device,
                &self.queue,
                &mut self.font_system,
                &mut self.atlas,
                Resolution {
                    width: self.size.width,
                    height: self.size.height,
                },
                [TextArea {
                    buffer: &self.buffer,
                    left: x as f32,
                    top: y as f32,
                    scale: 1.0,
                    bounds: TextBounds {
                        left: x as i32,
                        top: y as i32,
                        right: self.size.width as i32,
                        bottom: self.size.height as i32,
                    },
                    default_color: glyphon::Color::rgb(0, 0, 0),
                }],
                &mut self.cache,
            )
            .unwrap();
```
But that does it with a offset. Now i would need something like the worlds famous translate(-50%) css trick...
But I can't seem to find any API to get the size of the text. Also the `TextBounds` API is kinda weird because provides extra left and top values which are absolute so I need to pass the left and top in there as well...

**Comment by grovesNL** [comment]
The plan was to let cosmic-text handle all alignment, e.g., through https://docs.rs/cosmic-text/latest/cosmic_text/struct.BufferLine.html#method.set_align or any other APIs they add upstream. So the idea is that you could let the buffer be sized to the full width of the window and center it horizontally that way.

The extra left/top are useful if you want to render a region of the text area, so in those cases the left/top might not be the same (area left/top are about shifting, while bounds left/top are about clipping/culling). We might be able to make this a little easier though.

**Comment by DasLixou** [comment]
@grovesNL yeah from my perspective a more intuitive API would be to handle the left and top completely from the "align" sector and let the bounds be relative to the text and also provide a const `Bounds::FULL`

**Comment by DasLixou** [comment]
@grovesNL Is there already a way to use the alignment or is it like from a PR?

**Comment by grovesNL** [comment]
>Is there already a way to use the alignment or is it like from a PR?

I think you should already be able to call this function on your `Buffer`s.

> only provides basic left, middle, right alignment? No top, middle, down and no translating?

Vertical alignment was requested in https://github.com/pop-os/cosmic-text/issues/79 but hasn't been implemented yet as far as I know. Translating is handled by the left/top in the text area.

**Comment by DasLixou** [comment]
But then I would at least need an api to get the used size of the buffer

**Comment by grovesNL** [comment]
Yeah it's not very convenient right now, you could do something as described in https://github.com/pop-os/cosmic-text/issues/70 to get the visible lines

**Comment by grovesNL** [comment]
- I think buffer attachment sounds a bit too vague
- `horizontal_alignment` and `vertical_alignment` would be set on `buffer` itself, so we don't control that in glyphon and it can't be specified at this level
- I don't think I'd recommend using something like `Bounds::Unbound` because you almost always want to clip to some region (at least the window border) to avoid sending extra text to the GPU
- `base_color` sounds like we're using it to mix colors but it's more like a default vs. custom color for part of the text

**Comment by DasLixou** [comment]
> * I think buffer attachment sounds a bit too vague
> 
>     * `horizontal_alignment` and `vertical_alignment` would be set on `buffer` itself, so we don't control that in glyphon and it can't be specified at this level
> 
>     * I don't think I'd recommend using something like `Bounds::Unbound` because you almost always want to clip to some region (at least the window border) to avoid sending extra text to the GPU
> 
>     * `base_color` sounds like we're using it to mix colors but it's more like a default vs. custom color for part of the text

* I just was confused with all the Text... things so I named it that way, but when you think TextArea is more suitable, it's fine.
* So that would be something that happens at cosmic_text?
* Wait did I understand TextBounds wrong? I thought its at which point it does a linebreak, like a box in which the text is laid out. `Bounds::Unbound` should say: 'I never need a linebreak, just ignore when you are outside of the window'
* Oh yeah, base_color really is a bad choice with this in mind 😅

**Comment by grovesNL** [comment]
> So that would be something that happens at cosmic_text?

Exactly, most text styling, alignment, etc. all happen in cosmic-text itself instead of glyphon

> Wait did I understand TextBounds wrong? I thought its at which point it does a linebreak, like a box in which the text is laid out.

Text bounds is about clipping/culling to a visible area, the size of the buffer (the one you give to cosmic-text when you create a buffer) controls where line breaks happen

**Comment by DasLixou** [comment]
Wait... so when that's cosmic_texts stuff.. can we then not have the sizes of the text on render and use that to center it?

**Comment by grovesNL** [comment]
At least for vertically centering you'd have to query the height based on the visible lines which is https://github.com/pop-os/cosmic-text/issues/70

**Comment by DasLixou** [comment]
Why is this closed?

**Comment by DasLixou** [comment]
Well it's nothing on glyphons side but still it's not possible yet... so i think as a sort of tracking issue this will do it

**Comment by madmaxio** [comment]
Has someone tried implement cursor functionality with on top of cosmic-text/glyphon? Any recommendations what approach to use (cosmic-text alignment probably)?

**Comment by BillyDM** [comment]
> Has someone tried implement cursor functionality with on top of cosmic-text/glyphon? Any recommendations what approach to use (cosmic-text alignment probably)?

In the GUI lib I'm creating, I'm using the align inside of the text properties for horizontal alignment, and my own custom method for vertical alignment:

```rust
fn layout_text_bounds(
    bounds_size: Size,
    padding: Padding,
    min_clipped_size: Size,
    vertical_align: Align,
    font_size: f32,
    line_height: f32,
) -> Rect {
    let content_rect = crate::layout::layout_inner_rect_with_min_size(
        padding,
        Rect::from_size(bounds_size),
        min_clipped_size,
    );

    // We need to vertically align the text ourselves as rootvg/glyphon does not do this.
    let text_bounds_y = match vertical_align {
        Align::Start => content_rect.min_y(),
        Align::Center => content_rect.min_y() + ((content_rect.height() - font_size) / 2.0) + 1.0,
        Align::End => content_rect.max_y() - font_size,
    };

    Rect::new(
        Point::new(content_rect.min_x(), text_bounds_y),
        Size::new(content_rect.width(), line_height),
    )
}
```

**Comment by madmaxio** [comment]
> > Has someone tried implement cursor functionality with on top of cosmic-text/glyphon? Any recommendations what approach to use (cosmic-text alignment probably)?
> 
> In the GUI lib I'm creating, I'm using the align inside of the text properties for horizontal alignment, and my own custom method for vertical alignment:
> 
> ```rust
> pub fn layout_text_bounds(
>     bounds_size: Size,
>     unclipped_text_size: Size,
>     padding: Padding,
>     min_clipped_size: Size,
>     vertical_align: Align,
>     font_size: f32,
> ) -> Rect {
>     if unclipped_text_size.is_empty() {
>         return Rect::default();
>     }
> 
>     let content_rect = crate::layout::layout_inner_rect_with_min_size(
>         padding,
>         Rect::from_size(bounds_size),
>         min_clipped_size,
>     );
> 
>     // We need to vertically align the text ourselves as rootvg/glyphon does not do this.
>     let text_bounds_y = match vertical_align {
>         Align::Start => content_rect.min_y(),
>         Align::Center => content_rect.min_y() + ((content_rect.height() - font_size) / 2.0) + 1.0,
>         Align::End => content_rect.max_y() - unclipped_text_size.height,
>     };
> 
>     Rect::new(
>         Point::new(content_rect.min_x(), text_bounds_y),
>         Size::new(content_rect.width(), unclipped_text_size.height),
>     )
> }
> ```

Interesting, is it open source?

**Comment by BillyDM** [comment]
> Interesting, is it open source?

Yes, it is now. https://github.com/MeadowlarkDAW/Yarrow

**Comment by wick3dr0se** [comment]
You can checkout how we do it in egor as well: https://github.com/wick3dr0se/egor/blob/main/crates/egor_glue/src/text.rs
